src/threads/decimal.c

```diff
@@ -0,0 +1,91 @@
+#include <stdio.h>
+#include "decimal.h"
+
+#define SIGNED_INT_BITS_WITHOUT_SIGN 31
+#define DECIMAL_BYTES 17
+#define FRACTIONAL_BYTES (SIGNED_INT_BITS_WITHOUT_SIGN - DECIMAL_BYTES)
+
+static int f = 1 << FRACTIONAL_BYTES;
+static int MAX_SUPPORTED_VALUE = (1 << DECIMAL_BYTES) - 1;
+static int MIN_SUPPORTED_VALUE = -(1 << DECIMAL_BYTES);
+
+/*
+    https://pkuflyingpig.gitbook.io/pintos/appendix/4.4bsd-scheduler#fixed-point-summary
+*/
+decimal to_decimal(int integer)
+{
+    if (integer > MAX_SUPPORTED_VALUE || integer < MIN_SUPPORTED_VALUE)
+    {
+        // PANIC("Number is out of range of real arithmetic: '%i' ", integer);
+        decimal dec = {-1};
+        return dec;
+    }
+    decimal dec = {integer * f};
```

Comment:
Could we just do:
```
integer << 14
```
is it the same?

Reply:
Good question, in [the documentation there are 2 points](https://pkuflyingpig.gitbook.io/pintos/appendix/4.4bsd-scheduler#basics) why it makes sense to use multiplication instead of shifts:

> This section has consistently used multiplication or division by f, instead of q-bit shifts, for two reasons.
>* First, multiplication and division do not have the surprising operator precedence of the C shift operators.
>* Second, multiplication and division are well-defined on negative operands, but the C shift operators are not. Take care with these issues in your implementation.